types/react/index.d.ts

```diff
@@ -251,17 +220,17 @@ declare namespace React {
         ...children: ReactNode[]): DOMElement<P, T>;
 
     // Custom components
-    function cloneElement<P extends Q, Q>(
+    function cloneElement<P>(
         element: SFCElement<P>,
-        props?: Q, // should be Q & Attributes, but then Q is inferred as {}
+        props?: Partial<P> & Attributes,
```

Comment:
```ts
function cloneElement<P, K extends keyof P>(
  element: SFCElement<P>,
  props?: (Pick<P, K> | P) & Attributes,
  ...children: ReactNode[]
): SFCElement<P>
```

Similarly for the other overloads. This is similar to the `setState` trick to avoid accepting `undefined` unless `undefined` is actually an acceptable value for the prop.

Reply:
I also wonder if there's a way to get a better type for `...children`, but that's probably for another PR.

Reply:
What is "P extends React.HTMLAttributes<T>"? Could you please provide more details?

Reply:
We should really try to make this work with Pick. Partial is evil.

Reply:
I'd say it's better to get this in with `Partial` first, as it's still better than requiring an `as` cast. Figuring out how to get `Pick` working would be a future task... 😢